### PR TITLE
fix(train): improve quality of training

### PR DESCRIPTION
**Pull request by Lordmau5**
This PR includes several things discussed in #270 
- It ensures the `_temp_epoch` variable is initialized in the LightningModule's `__init__` method (thanks @ne0escape)
- It moves the `__init__` method of the LightningModule to the top (bit of cleanup)
- It fixes the order of optimization done to the model as per the Lightning.AI documentation ([lightning.ai/docs/pytorch/stable/common/optimization.html#id2](https://lightning.ai/docs/pytorch/stable/common/optimization.html#id2))
- It removes a `with torch.no_grad():` call for the generator loss calculation which ends up vastly improving the quality of the model (less metallic noise during several tests with a model from scratch)
- It fixes the `log_audio_dict` function so it use the correct `total_batch_idx` value
- It fixes checkpoints saving twice (or multiple times) since `validation_step` is being run for every model apparently (we check if `batch_idx` is 0 and only then save)

Please tell me if you'd like to see any changes made before potentially merging this 🙏

### CODE REVIEW

**Comment by Lordmau5** [comment]
Two examples of the TF2 Spy after only 200 iterations with the fix:
[original_1.webm](https://user-images.githubusercontent.com/1345036/230796583-11c5c93c-5c19-4f33-8349-851e5efcb43d.webm)
[generated_1_200.webm](https://user-images.githubusercontent.com/1345036/230796587-212c3fb8-f301-416d-9382-01dd688e4658.webm)

[original_2.webm](https://user-images.githubusercontent.com/1345036/230796590-e8036241-6548-4040-81f0-ed3769373576.webm)
[generated_2_200.webm](https://user-images.githubusercontent.com/1345036/230796594-4dafda25-4c3c-4bbd-b3b5-6f94091debad.webm)

**Comment by 34j** [comment]
😍🥰😎
